cpp/tests/stream_compaction/distinct_count_tests.cpp

```diff
@@ -0,0 +1,367 @@
+}
+
+TYPED_TEST(DistinctCountCommon, TableNoNull)
+{
+  using T = TypeParam;
+
+  auto const input1 = cudf::test::make_type_param_vector<T>(
+    {1, 3, 3, 3, 4, 31, 1, 8, 2, 0, 4, 1, 4, 10, 40, 31, 42, 0, 42, 8, 5, 4});
+  auto const input2 = cudf::test::make_type_param_vector<T>(
+    {3, 3, 3, 4, 31, 1, 8, 5, 0, 4, 1, 4, 10, 40, 31, 42, 0, 42, 8, 5, 4, 1});
+
+  std::vector<std::pair<T, T>> pair_input;
+  std::transform(
+    input1.begin(), input1.end(), input2.begin(), std::back_inserter(pair_input), [](T a, T b) {
+      return std::make_pair(a, b);
+    });
+
+  cudf::test::fixed_width_column_wrapper<T> input_col1(input1.begin(), input1.end());
+  cudf::test::fixed_width_column_wrapper<T> input_col2(input2.begin(), input2.end());
+
+  std::vector<cudf::column_view> cols{input_col1, input_col2};
+  cudf::table_view input_table(cols);
+
+  cudf::size_type const expected =
+    std::set<std::pair<T, T>>(pair_input.begin(), pair_input.end()).size();
+  EXPECT_EQ(expected, cudf::unordered_distinct_count(input_table, null_equality::EQUAL));
+
+  auto const new_end      = std::unique(pair_input.begin(), pair_input.end());
+  auto const gold_ordered = new_end - pair_input.begin();
+  EXPECT_EQ(gold_ordered, cudf::distinct_count(input_table, null_equality::EQUAL));
+}
+
+struct DistinctCount : public cudf::test::BaseFixture {
+};
+
+TEST_F(DistinctCount, WithNull)
+{
+  using T = int32_t;
+
+  std::vector<T> input               = {1,   3,  3,  XXX, 31, 1, 8,  2, 0, XXX, XXX,
+                          XXX, 10, 40, 31,  42, 0, 42, 8, 5, XXX};
+  std::vector<cudf::size_type> valid = {1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0,
+                                        0, 1, 1, 1, 1, 1, 1, 1, 1, 0};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col(input.begin(), input.end(), valid.begin());
+
+  // explicit instantiation to one particular type (`double`) to reduce build time
+  cudf::size_type const expected = std::set<double>(input.begin(), input.end()).size();
+  EXPECT_EQ(
+    expected,
+    cudf::unordered_distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_VALID));
+
+  auto const new_end      = std::unique(input.begin(), input.end());
+  auto const gold_ordered = new_end - input.begin() - 3;
+  EXPECT_EQ(gold_ordered,
+            cudf::distinct_count(input_col, null_policy::EXCLUDE, nan_policy::NAN_IS_VALID));
+}
+
+TEST_F(DistinctCount, IgnoringNull)
+{
+  using T = int32_t;
+
+  std::vector<T> input               = {1,   YYY, YYY, XXX, 31, 1, 8,  2, 0, XXX, 1,
+                          XXX, 10,  40,  31,  42, 0, 42, 8, 5, XXX};
+  std::vector<cudf::size_type> valid = {1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1,
+                                        0, 1, 1, 1, 1, 1, 1, 1, 1, 0};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col(input.begin(), input.end(), valid.begin());
+
+  cudf::size_type const expected = std::set<T>(input.begin(), input.end()).size();
+  // Removing 2 from expected to remove count for `XXX` and `YYY`
```

Comment:
This is from the existing test. I've updated the tests a bit so two null value masks are meaningful for the `detail::distinct_count` test.

Reply:
I was referring to the test where the sequence of {`YYY`, `YYY`, `XXX`} is considered as one group of equivalent rows. So in the `detail::distinct_count` test right after: https://github.com/rapidsai/cudf/pull/10030/files/a1514436c74ff9cb34e852fb8aa4816c87fdc451#diff-c87539239022212c325d533c8a9470cec4b89f5fb999f607acd1ecf1271191f5R142-R146, we need to subtract 5 instead of 4 since {`YYY`, `YYY`, `XXX`} is treated as two distinct groups in `std::unque`. By doing this, we also verified if the implicit "nulls are equal" is properly handled.

Reply:
> By doing this, we also verified if the implicit "nulls are equal" is properly handled.

I see why this might be helpful, but it does seem to be testing an equality behavior that should be guaranteed by, and tested by, a lower level of the code (specifically the `row_equality_comparator`). I hope we shouldn't need to explicitly test that two different values that each have a null bit in the mask are equivalent in every higher-level algorithm relying on that behavior...

Reply:
It's not only the row equality. If you remember, we have [several if statements](https://github.com/PointKernel/cudf/blob/optimize-compaction/cpp/src/stream_compaction/distinct_count.cu#L223-L233) around NaN and null so I would like to have those tested as well.

Reply:
I don't see why that code would merit having tests for the explicit case of two different values that are both marked as null... the case I think you'd need to watch out for is having two equivalent values with different nullity, e.g. `[42, 42]` with mask `[0, 1]`. We can resolve this if you want to keep the current tests - there are lots of other conversations on this PR and I don't want this to be a blocker.